keps/sig-auth/3221-structured-authorization-configuration/README.md

```diff
@@ -0,0 +1,775 @@
+respectively.
+
+Today, the `SubjectAccessReview` version defaults to `v1beta1` if the corresponding
+flag is not supplied. While configuring authorization modes using the file config,
+the version supported by a webhook has to be mentioned using a required field
+`subjectAccessReviewVersion`.
+
+The user can define a CEL expression to determine whether a request needs to dispatched
+to the authz webhook for which the expression has been defined. The user would have access
+to a `request` variable containing a `SubjectAccessReview` object in the version specified
+by `subjectAccessReviewVersion`.
+
+The code path for enabling the above will only be triggered if the feature flag is enabled until 
+the feature flag is removed and this feature graduates to GA.
+
+### Monitoring
+
+We will add the following 4 metrics:
+
+1. `apiserver_authorization_step_invocations_total`
+
+This will be incremented on round-trip of an authorizer. It will track total
+authorization decision invocations across the following labels.
+
+Labels {along with possible values}:
+- `mode` {<authorizer_name>}
+- `decision` {Allow, Deny, NoOpinion}
+
+**Note:** Some examples of <authorizer_name>: `RBAC`, `Node`, `ABAC`, `webhook{,_<name>}`.
+If there is only one webhook, there would be no `_<name>` suffix.
+
+2. `apiserver_authorization_step_webhook_invocations_total`
+
+This will be incremented on round-trip of an authorization webhook. It will track
+total invocation counts across the following labels.
+
+- `name`
+- `code` {2xx, 4xx, 5xx}
+- `decision` {Allow, Deny, NoOpinion}
+
+3. `apiserver_authorization_step_webhook_duration_total_seconds`
+
+This metric will track the total round trip time of the requests to the webhook.
+
+Labels {along with possible values}:
+- `name`
+- `code` {2xx, 4xx, 5xx}
+- `decision` {Allow, Deny, NoOpinion}
+
+4. `apiserver_authorization_step_webhook_error_total`
+
+This metric will be incremented when a webhook returns a 4xx or 5xx (erroneous) response.
+
+Labels {along with possible values}:
+
+- `name`
+- `code` {4xx, 5xx}
+- `decision` {Deny, NoOpinion}
+
+5. `apiserver_authorization_step_configuration_reload_failure_count`
+
+This metric would track the number of times the configuration has been reloaded.
+
+### Test Plan
+
+[x] I/we understand the owners of the involved components may require updates to
+existing tests to make this code solid enough prior to committing the changes necessary
+to implement this enhancement.
+
+##### Prerequisite testing updates
+
+##### Unit tests
+
+Unit tests would be added along with any new code introduced.
+
+Existing test coverage:
+- cmd/kube-apiserver/app/options: 65
+- staging/src/k8s.io/apiserver/pkg/authorization: NA
+- staging/src/k8s.io/apiserver/apis/config/v1: 54.5
+
+> Note: We will add more packages to the list during implementation.
+
+**Benchmarks**
+
+We should benchmark the cost of some common CEL expressions inside
```

Comment:
@jpbetz we should do this for CEL in general since this is not specific to this KEP only. WDYT?

Reply:
xref: https://github.com/google/cel-go/issues/729

Reply:
Ensuring that CEL performance doesn't regress is something I'd like to see owned primarily by cel-go. I suspect there are already things in place. I added a comment on https://github.com/google/cel-go/issues/729 to see if anything can be shared.

That still leaves room for benchmarking on the k8s side. The [CEL Cost model](https://docs.google.com/spreadsheets/d/123G4mLDTWjpCCbfop7HL3NkDvIBLaPu6YqV8ro3Isnw/edit#gid=1510432316) data might be useful in some contexts.

I think it is reasonable for features like proposed by this KEP to add some go mincor benchmarks a cases you care about. This is a relatively low effort thing to do. Typically when people see a PR that they're worried will impact a micro benchmark they ask the PR author to paste before/after output into the PR description.

We do have line items to better track performance before we take features to GA.  I think we'll spend some time in 1.29 on it. It might be that we rely more on metrics than on micro benchmarks?  I'm not sure yet.